parsely/src/main/java/com/parsely/parselyandroid/ParselyTracker.kt

```diff
@@ -0,0 +1,424 @@
+        )
+        intervalCalculator = HeartbeatIntervalCalculator(clock)
+
+        isDebug = false
+        flushManager.start()
+        ProcessLifecycleOwner.get().lifecycle.addObserver(
+            LifecycleEventObserver { _, event: Lifecycle.Event ->
+                if (event == Lifecycle.Event.ON_STOP) {
+                    flushEvents()
+                }
+            }
+        )
+    }
+
+    /**
+     * Get the heartbeat interval
+     *
+     * @return The base engagement tracking interval.
+     */
+    public val engagementInterval: Double?
+        get() = engagementManager?.intervalMillis
+
+    public val videoEngagementInterval: Double?
+        get() = videoEngagementManager?.intervalMillis
+
+    /**
+     * Returns whether the engagement tracker is running.
+     *
+     * @return Whether the engagement tracker is running.
+     */
+    public fun engagementIsActive(): Boolean {
+        return engagementManager?.isRunning ?: false
+    }
+
+    /**
+     * Returns whether video tracking is active.
+     *
+     * @return Whether video tracking is active.
+     */
+    public fun videoIsActive(): Boolean {
+        return videoEngagementManager?.isRunning ?: false
+    }
+
+    /**
+     * Returns the interval at which the event queue is flushed to Parse.ly.
+     *
+     * @return The interval at which the event queue is flushed to Parse.ly.
+     */
+    public val flushInterval: Long
+        get() = flushManager.intervalMillis / 1000
+
+    /**
+     * Set a debug flag which will prevent data from being sent to Parse.ly
+     *
+     *
+     * Use this flag when developing to prevent the SDK from actually sending requests
+     * to Parse.ly servers. The value it would otherwise send is logged to the console.
+     *
+     * @param debug Value to use for debug flag.
+     */
+    public fun setDebug(debug: Boolean) {
+        isDebug = debug
+        log("Debugging is now set to $isDebug")
+    }
+
+    /**
+     * Register a pageview event using a URL and optional metadata.
+     *
+     * @param url         The URL of the article being tracked
+     * (eg: "http://example.com/some-old/article.html")
+     * @param urlRef      Referrer URL associated with this video view.
+     * @param urlMetadata Optional metadata for the URL -- not used in most cases. Only needed
+     * when `url` isn't accessible over the Internet (i.e. app-only
+     * content). Do not use this for **content also hosted on** URLs Parse.ly
+     * would normally crawl.
+     * @param extraData   A Map of additional information to send with the event.
+     */
+    public fun trackPageview(
+        url: String,
+        urlRef: String? = null,
+        urlMetadata: ParselyMetadata? = null,
+        extraData: Map<String, Any>? = null,
+    ) {
+        if (url.isBlank()) {
+            log("url cannot be empty")
+            return
+        }
+
+        val pageViewUuid = generatePixelId()
+        lastPageviewUuid = pageViewUuid
```

Comment:
Nitpick:

```suggestion
        lastPageviewUuid = generatePixelId()
```

I think we can drop the extra value.

Reply:
We could, but the Kotlin compiler can't infer that `lastPageviewUuid` is not null. I think this additional variable fits better than `!!` - it always feels like a code smell, and was designed to feel like one. WDYT?

<img width="1029" alt="image" src="https://github.com/Parsely/parsely-android/assets/5845095/8a901b3c-732b-40e6-b2ad-b8386eef4b36">

Reply:
That's a good point. I completely missed the mutability and this was a pretty bad suggestion. Sorry about that!